src/mygrad/nnet/activations/softmax.py

```diff
@@ -85,54 +85,50 @@ def softmax(x, constant=False):
     Tensor([[0.33333333, 0.33333333, 0.33333333],
             [0.5       , 0.5       , 0.        ]])
     """
-    return Tensor._op(Softmax, x, constant=constant)
+    return Tensor._op(Softmax, x, op_kwargs=dict(axis=axis), constant=constant)
 
 
 class LogSoftmax(Operation):
     scalar_only = True
 
-    def __call__(self, a):
+    def __call__(self, a, axis=-1):
         self.variables = (a,)
         x = a.data
-        assert 0 < a.ndim < 3
 
-        self.__kw = (
-            dict(axis=1, keepdims=True)
-            if x.ndim == 2
-            else dict(axis=None, keepdims=False)
-        )
-        return x - _logsumexp(x, **self.__kw)
+        self._kw = dict(axis=axis, keepdims=True)
+        return x - _logsumexp(x, **self._kw)
 
     def backward_var(self, grad, index, **kwargs):
         a = self.variables[index]
         x = a.data
-        soft = _softmax(x, self.__kw)
-        return grad - soft * np.sum(grad, **self.__kw)
+        soft = _softmax(x, self._kw)
+        return grad - soft * np.sum(grad, **self._kw)
 
 
-def logsoftmax(x, constant=False):
+def logsoftmax(x, axis=-1, constant=False):
     r"""
     Applies the log-softmax activation function::
 
         f(x) = log ( exp(x) / sum( exp(x) ) )
 
-    Compute the softmax over a 1D tensor of data, or along the
-    respective rows of a 2D tensor
+    Computes the log-softmax over one or more axes of an ND-tensor.
 
     Parameters
     ----------
-    x : array_like, shape=(D,) or shape=(N,D)
-        log-softmax is computed along the rows of ``x`` if
-        ``x`` is a 2D array. Otherwise log-softmax is computed
-        on the 1D ``x``.
+    x : array_like
+
+    axis : Union[None, int, Tuple[int, ...]], optional (default=-1)
```

Comment:
for these can we specify the `axis` as a `mg.Tensor` or `np.ndarray`? do we want to allow that?

Reply:
At that point I might as well put `Sequence[int]`. I think I'll just keep it simple though